plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockStorage.java

```diff
@@ -19,289 +19,565 @@
-import com.google.api.services.storage.Storage;
-import com.google.api.services.storage.model.Bucket;
-import com.google.api.services.storage.model.StorageObject;
-import org.elasticsearch.common.io.Streams;
-import org.elasticsearch.rest.RestStatus;
-
-import java.io.ByteArrayInputStream;
+import org.elasticsearch.common.SuppressForbidden;
+import org.mockito.Matchers;
+
+import com.google.api.gax.paging.Page;
+import com.google.cloud.Policy;
+import com.google.cloud.ReadChannel;
+import com.google.cloud.RestorableState;
+import com.google.cloud.WriteChannel;
+import com.google.cloud.storage.Acl;
+import com.google.cloud.storage.Acl.Entity;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.CopyWriter;
+import com.google.cloud.storage.ServiceAccount;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.spi.v1.StorageRpc;
+import com.google.cloud.storage.StorageBatch;
+import com.google.cloud.storage.StorageOptions;
+
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.math.BigInteger;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.nio.ByteBuffer;
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
 
+import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 /**
  * {@link MockStorage} mocks a {@link Storage} client by storing all the blobs
  * in a given concurrent map.
  */
-class MockStorage extends Storage {
+class MockStorage implements Storage {
+
+    private final Bucket theBucket;
+    private final ConcurrentMap<String, Blob> blobsMap;
+
+    @SuppressForbidden(reason = "mocking here requires reflection that trespasses the access system")
+    MockStorage(final String bucketName, final ConcurrentMap<String, Blob> blobsMap) {
+        this.blobsMap = blobsMap;
+        // mock bucket
+        this.theBucket = mock(Bucket.class);
+        when(this.theBucket.getName()).thenReturn(bucketName);
+        doAnswer(invocation -> {
+            assert invocation.getArguments().length == 1 : "Only a single filter is mocked";
+            final BlobListOption prefixFilter = (BlobListOption) invocation.getArguments()[0];
+            final Method optionMethod = BlobListOption.class.getSuperclass().getDeclaredMethod("getRpcOption");
+            optionMethod.setAccessible(true);
+            assert StorageRpc.Option.PREFIX.equals(optionMethod.invoke(prefixFilter)) : "Only the prefix filter is mocked";
+            final Method valueMethod = BlobListOption.class.getSuperclass().getDeclaredMethod("getValue");
+            valueMethod.setAccessible(true);
```

Comment:
These, unfortunately require the `suppressAccessChecks` permission. (Answer to review comment: https://github.com/elastic/elasticsearch/pull/30168/files#r184983763)

Reply:
@albertzaharovits and I talked about this, and we figured out that this is not the only place where the permission is needed but the main code requires it too.

I still think that we should not add a specific permission for tests in this case because it can hide other places where a security exception could be thrown where in fact we should work to remove this permission in the main plugin. I also think that in this case Mockito makes things harder to maintain and to read where we can instead [use POJOs to mock the Google Storage](https://github.com/tlrx/elasticsearch/blob/1f40ea509406b14ac4300828b27b9418562eee59/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockStorage.java) but with this option we have to pay the price of adding two utility classed in foreign packages like `com.google.storage`. It's like to choose between two bad options but this is an interesting point and I'd like to know @elastic/es-core-infra opinion

Reply:
Anything is better than allowing/using setAccessible. Creating utility classes in a foreign package is not great, but definitely preferable, IMO.

Reply:
Thanks @rjernst and core/infra team.